lib/circuit-breaker.ts

```diff
@@ -0,0 +1,233 @@
+  volumeThreshold?: number;
+
+  onCircuitOpen?: (m: Metrics) => void;
+  onCircuitClose?: (m: Metrics) => void;
+}
+
+const enum State {
+  OPEN,
+  HALF_OPEN,
+  CLOSED
+}
+
+export interface Metrics {
+  totalCount: number;
+  errorCount: number;
+  errorPercentage: number;
+}
+
+interface Bucket {
+  failures: number;
+  successes: number;
+  timeouts: number;
+  shortCircuits: number;
+}
+
+export type Command = (success: () => void, failure: () => void) => void;
+
+function noop() {}
+
+export interface CircuitBreakerPublicApi {
+  run(command: Command, fallback?: () => void): void;
+  forceClose(): void;
+  forceOpen(): void;
+  unforce(): void;
+  isOpen(): boolean;
+}
+
+export class CircuitBreaker implements CircuitBreakerPublicApi {
+  public windowDuration: number;
+  public timeoutDuration: number;
+  public errorThreshold: number;
+  public volumeThreshold: number;
+
+  public onCircuitOpen: (m: Metrics) => void;
+  public onCircuitClose: (m: Metrics) => void;
+
+  private readonly buckets: Bucket[];
+  private bucketIndex: number;
+  private state: State;
+  private forced?: State;
+
+  constructor(options?: CircuitBreakerOptions) {
+    options = options || {};
+
+    this.windowDuration = options.windowDuration || 10000;
+    this.timeoutDuration = options.timeoutDuration || 3000;
+    this.errorThreshold = options.errorThreshold || 50;
+    this.volumeThreshold = options.volumeThreshold || 5;
+
+    this.onCircuitOpen = options.onCircuitOpen || noop;
+    this.onCircuitClose = options.onCircuitClose || noop;
+
+    this.buckets = [];
+    const numberOfBuckets = Math.max(1, options.numBuckets || 10);
+    for (let i = 0; i < numberOfBuckets; ++i) {
+      this.buckets.push({
+        failures: 0,
+        successes: 0,
+        timeouts: 0,
+        shortCircuits: 0
+      });
+    }
+    this.bucketIndex = 0;
+    this.state = State.CLOSED;
+    this.forced = undefined;
+
+    this.startTicker();
+  }
+
+  public run(command: Command, fallback?: () => void) {
+    if (this.isOpen()) {
+      this.executeFallback(fallback || function() {});
+    } else {
+      this.executeCommand(command);
+    }
+  }
+
+  public forceClose() {
+    this.forced = this.state;
+    this.state = State.CLOSED;
+  }
+
+  public forceOpen() {
+    this.forced = this.state;
+    this.state = State.OPEN;
+  }
+
+  public unforce() {
+    if (this.forced !== undefined) {
+      this.state = this.forced;
+      this.forced = undefined;
+    }
+  }
+
+  public isOpen() {
+    return this.state === State.OPEN;
+  }
+
+  private startTicker() {
+    const self = this;
+    const bucketDuration = this.windowDuration / this.buckets.length;
+
+    const tick = () => {
+      ++this.bucketIndex;
+
+      // FIXME this is very broken as it means that the time
+      //       till CB is changing to half-open state depends on
+      //       the index of the bucket at the time when it opened.
+      if (this.bucketIndex >= this.buckets.length) {
```

Comment:
This is how it was in the original code. Fixing it will mean changing CB behavior and will be tackled separately.